processor/metricstarttimeprocessor/internal/subtractinitial/adjuster.go

```diff
@@ -19,19 +22,354 @@ import (
+	for i := 0; i < currentPoints.Len(); i++ {
+		currentDist := currentPoints.At(i)
+
+		referenceTsi, found := referenceTsm.Get(current, currentDist.Attributes())
+		if !found {
+			// initialize everything. Don't add the datapoint to the result.
+			referenceTsi.ExponentialHistogram = pmetric.NewExponentialHistogramDataPoint()
+			currentDist.CopyTo(referenceTsi.ExponentialHistogram)
+			continue
+		}
+
+		// Adjust the datapoint based on the reference value.
+		adjustedPoint := pmetric.NewExponentialHistogramDataPoint()
+		currentDist.CopyTo(adjustedPoint)
+		adjustedPoint.SetStartTimestamp(referenceTsi.ExponentialHistogram.StartTimestamp())
+		if currentDist.Flags().NoRecordedValue() {
+			// TODO: Investigate why this does not reset.
+			tmp := resExpHistogram.DataPoints().AppendEmpty()
+			adjustedPoint.CopyTo(tmp)
+			continue
+		}
+		subtractExponentialHistogramDataPoint(adjustedPoint, referenceTsi.ExponentialHistogram)
+
+		previousTsi, found := previousValueTsm.Get(current, currentDist.Attributes())
+		if !found {
+			// First point after the reference. Not a reset.
+			previousTsi.ExponentialHistogram = pmetric.NewExponentialHistogramDataPoint()
+		} else if previousTsi.IsResetExponentialHistogram(adjustedPoint) {
+			// reset re-initialize everything and use the non adjusted points start time.
+			referenceTsi.ExponentialHistogram = pmetric.NewExponentialHistogramDataPoint()
+			referenceTsi.ExponentialHistogram.SetStartTimestamp(currentDist.StartTimestamp())
+
+			tmp := resExpHistogram.DataPoints().AppendEmpty()
+			currentDist.CopyTo(tmp)
+			continue
+		}
+
+		// Update previous values with the current point.
+		adjustedPoint.CopyTo(previousTsi.ExponentialHistogram)
+		tmp := resExpHistogram.DataPoints().AppendEmpty()
+		adjustedPoint.CopyTo(tmp)
+	}
+}
+
+func adjustMetricSum(referenceTsm, previousValueTsm *datapointstorage.TimeseriesMap, current pmetric.Metric, resSum pmetric.Sum) {
+	resSum.SetAggregationTemporality(current.Sum().AggregationTemporality())
+	resSum.SetIsMonotonic(current.Sum().IsMonotonic())
+
+	currentPoints := current.Sum().DataPoints()
+	for i := 0; i < currentPoints.Len(); i++ {
+		currentDist := currentPoints.At(i)
+
+		referenceTsi, found := referenceTsm.Get(current, currentDist.Attributes())
+		if !found {
+			// initialize everything. Don't add the datapoint to the result.
+			referenceTsi.Number = pmetric.NewNumberDataPoint()
+			currentDist.CopyTo(referenceTsi.Number)
+			continue
+		}
+
+		// Adjust the datapoint based on the reference value.
+		adjustedPoint := pmetric.NewNumberDataPoint()
+		currentDist.CopyTo(adjustedPoint)
+		adjustedPoint.SetStartTimestamp(referenceTsi.Number.StartTimestamp())
+		if currentDist.Flags().NoRecordedValue() {
+			// TODO: Investigate why this does not reset.
+			tmp := resSum.DataPoints().AppendEmpty()
+			adjustedPoint.CopyTo(tmp)
+			continue
+		}
+		adjustedPoint.SetDoubleValue(adjustedPoint.DoubleValue() - referenceTsi.Number.DoubleValue())
```

Comment:
I feel this might be wrong and might be the root casue fo the unexpected case https://github.com/bwplotka/otel-ops-col/pull/1

I think we need to first check `previousValueTsm`, do the `reviousTsi.IsResetSum(NOTadjustedPoint)` on not adjusted points, otherwise we might miss reset because of the normalization.

Reply:
This was a great catch actually and this flow is very confusing. I added some more testing to make sure this doesn't happen and I now do check against the reference cache too if the value decreased. Thanks!

Test added that tests this: `TestResetAfterInitialStart`